### PR TITLE
Microwave upgrades: Microwave PDA charging, power cells

### PR DESCRIPTION
**Pull request by lessthnthree**
## About The Pull Request

For the hard-working tradesperson who's in the middle of nowhere and just wants to warm up their pastry-based savoury item purchased from an overpriced vending machine, introduces the engineer themed wireless microwave.

No more going hungry in a pinch, this variant uses changeable power cells instead of APC power. Can be built normally or printed with a furnishings-upgraded RCD.

The other upgrade now available: PDA charging. Upgrade the capacitors in the microwave to add a Charge mode for your portable device!

Also cleans up microwave code a bit and adds much needed context hints for when it's broken, dirty, etc.
Swaps a few break room microwaves to upgraded variants on mapload, adds a microwave/donks to Birdshot, donks to Tramstation.

## Why It's Good For The Game

Slap together a semi-portable microwave with your components on hand, have an engineering picnic in space or wherever else the day takes you.

With increased PDA drain, adding another option to charge your PDA in common areas is useful. Microwaves are already strategically placed around the station. Please ensure the model you're using has the charging feature before operating.

image
image

## Changelog

:cl: LT3
add: Introducing Nanotrasen Wave! A Nanotrasen exclusive, Waveallows your PDA to be charged wirelessly through microwave frequencies. You can Wave-charge your device by placing it inside a compatible microwave and selecting the charge mode.
add: Microwaves can be upgraded to add wireless charging
add: Cell-swappable microwave for the engineer on-the-go
add: Microwave now has a wire to swap charge/cook modes
add: Furnishings RCD upgrade now includes wireless microwave
add: Tramstation and Birdshot engineering break rooms now have microwave and donk pockets. Some microwaves come pre-equipped with wireless charging and an upgraded cell.
fix: The microwave in the snowdin ruin is now real, not a fluff prop
/:cl:

### CODE REVIEW

**Comment by Loafin34** [comment]
sovl... removed...

**Comment by SomeoneYouProbablyKnow** [comment]
Now we merge this and then silently revert it in a month 😈

**Comment by Wallemations** [comment]
Also wait, why does the donk pocket not look like it has an overlay on it?

image

It just looks like it's kinda floating, makes it look really weird compared to normal microwaves.

**Comment by CRITAWAKETS** [comment]
it'd be funnier if cook mode was what you pressed to charge and charge functionality was only told on examine, so there's a slight amount of risk if you're inattentive

**Comment by lessthnthree** [comment]
> it'd be funnier if cook mode was what you pressed to charge and charge functionality was only told on examine, so there's a slight amount of risk if you're inattentive

Don't worry, the right-click shortcut to start the microwave is still present. If you opt to go for the shortcut, you do however need to examine to know which mode it's currently set to.

**Comment by lessthnthree** [comment]
> Also wait, why does the donk pocket not look like it has an overlay on it?
> 
> image
> 
> It just looks like it's kinda floating, makes it look really weird compared to normal microwaves.
> 
> Also, do you think you could try making the greys a little darker? I think the increased contrast between the yellow and grey would make it look better.

Sure. I think I made the door overlay too transparent. I'll hit you up in discord with adjusted colors.

**Comment by iwishforducks** [comment]
lame tbh. keep pda microwaving charging out of code. it’s even funnier tricking people when there is straight up no way to charge using microwaves

**Comment by HalfPunch** [comment]
image

**Comment by lessthnthree** [comment]
> Also wait, why does the donk pocket not look like it has an overlay on it?
> 
> image
> 
> It just looks like it's kinda floating, makes it look really weird compared to normal microwaves.
> 
> Also, do you think you could try making the greys a little darker? I think the increased contrast between the yellow and grey would make it look better.

Fixed the overlays, darkened the engineering variant.

image

**Comment by lessthnthree** [comment]
> image

image

**Review by Wallemations — APPROVED** [review]
I'd look into making that big blue light match the rest of the palette but other than that it's good

**Comment by lessthnthree** [comment]
> I'd look into making that big blue light match the rest of the palette but other than that it's good

Yeah it's a bit bright. I'll throw it in a nogbp update.

**Comment by Wallemations** [comment]
> > I'd look into making that big blue light match the rest of the palette but other than that it's good
> 
> Yeah it's a bit bright. I'll throw it in a nogbp update.

why wait instead of just doing it now lmao
that's just giving yourself way more work instead of updating one sprite

**Comment by lessthnthree** [comment]
> > > I'd look into making that big blue light match the rest of the palette but other than that it's good
> > 
> > 
> > Yeah it's a bit bright. I'll throw it in a nogbp update.
> 
> why wait instead of just doing it now lmao that's just giving yourself way more work instead of updating one sprite

wasn't sure when I'd have time lol.

anyways done, the light is reduced and matching the rest

image

File: code/game/objects/items/circuitboards/machines/machine_circuitboards.dm [hunk 1]
```diff
@@ -1245,10 +1245,23 @@
...
+	name = "Wireless Microwave Oven"
+	greyscale_colors = CIRCUIT_COLOR_SERVICE
+	build_path = /obj/machinery/microwave/engineering
+	req_components = list(
+		/datum/stock_part/micro_laser = 1,
+		/datum/stock_part/matter_bin = 1,
+		/datum/stock_part/capacitor/tier2 = 1,
+		/obj/item/stack/cable_coil = 4,
+		/obj/item/stack/sheet/glass = 2)
+	needs_anchored = FALSE
```
Comment (Ghommie) [inline code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:1263]:
```suggestion
```
same value as parent type.

File: code/modules/food_and_drinks/machinery/microwave.dm [hunk 2]
```diff
@@ -37,20 +44,46 @@
...
 	var/efficiency = 0
+	/// If we use a cell instead of powernet
+	var/cell_powered = FALSE
+	/// The cell we charge with
+	var/obj/item/stock_parts/cell/cell
+	/// The cell we're charging
+	var/obj/item/stock_parts/cell/vampire_cell
+	/// Capable of vampire charging PDAs
+	var/vampire_charging_capable = FALSE
+	var/vampire_charging_enabled = FALSE
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/microwave.dm:55]: documentation. I was about to confuse one with the other too.

File: code/modules/food_and_drinks/machinery/microwave.dm [hunk 3]
```diff
@@ -106,44 +186,40 @@
...
+		"[span_notice("- Mode: <b>[vampire_charging_enabled ? "Charge" : "Cook"]</b>.")]\n"+\
 		"[span_notice("- Capacity: <b>[max_n_of_items]</b> items.")]\n"+\
-		span_notice("- Cook time reduced by <b>[(efficiency - 1) * 25]%</b>.")
+		span_notice("- Power: <b>[efficiency * TIER_1_CELL_CHARGE_RATE]W</b>.")
+
+		if(cell_powered)
+			if(!isnull(cell))
+				. += span_notice("- Charge: <b>[round(cell.percent())]%</b>.")
+			else
+				. += span_notice("- Charge: <b>INSERT CELL</b>.")
```
Comment (Ghommie) [inline code/modules/food_and_drinks/machinery/microwave.dm:209]:
```suggestion
			. += span_notice("- Charge: <b>[isnull(cell) ? "INSERT CELL" : "[round(cell.percent())]%"</b>.")
```

**Review by Ghommie — CHANGES_REQUESTED** [review]
Remember kids, do not try this at home.

File: code/modules/food_and_drinks/machinery/microwave.dm (hunk 2 above)
Comment (lessthnthree) [reply code/modules/food_and_drinks/machinery/microwave.dm:55]: Ah yes, of all the vars that should have been doc'd, I missed the important ones

**Review by Ghommie — APPROVED** [review]
Nothing glaring at a second glance.

File: code/game/objects/structures/showcase.dm [hunk 4]
```diff
@@ -83,7 +83,13 @@
...
 	desc = "The famous Nanotrasen-brand microwave, the multi-purpose cooking appliance every station needs! This one appears to be drawn onto a cardboard box."
 	icon = 'icons/obj/machines/microwave.dmi'
-	icon_state = "map_icon"
+	icon_state = "mw_complete"
+
+/obj/structure/showcase/machinery/microwave_engineering
+	name = "\improper Nanotrasen Wave(tm) microwave"
+	desc = "Just when everyone thought Nanotrasen couldn't improve on their famous microwave, they added wireless induction charging for your PDA! Groundbreaking!"
+	icon = 'icons/obj/machines/microwave.dmi'
+	icon_state = "engi_mw_complete"
```
Comment (Ghommie) [inline code/game/objects/structures/showcase.dm:92]: Actually, is this showcase even used?

Comment (Ghommie) [reply code/game/objects/structures/showcase.dm:92]: But it doesn't look like your PR is mapping it in.

Comment (lessthnthree) [reply code/game/objects/structures/showcase.dm:92]: I just went and confirmed, the obj itself doesn't need to be mapped. The corporate showcase has displays where it spawns a random /obj/structure/showcase object.

Comment (lessthnthree) [reply code/game/objects/structures/showcase.dm:92]: Oh, I see it now.

**Pull request merged**